src/datum_api.c

```diff
@@ -491,7 +492,8 @@ bool datum_api_check_admin_password(struct MHD_Connection * const connection, co
 		datum_api_submit_uncached_response(connection, MHD_HTTP_FORBIDDEN, auth_failure_response_creator());
 		return false;
 	}
-	if (!datum_secure_strequals(datum_config.api_csrf_token, sizeof(datum_config.api_csrf_token)-1, json_string_value(j_csrf))) {
+	const char * const csrf_s = json_string_value(j_csrf);
+	if ((strlen(csrf_s) == sizeof(datum_config.api_csrf_token)-1) && 0 != sodium_memcmp(datum_config.api_csrf_token, json_string_value(j_csrf), sizeof(datum_config.api_csrf_token)-1)) {
```

Comment:
The length check before calling sodium_memcmp should not leak any information because the length of CSRF tokens is public information (64 bytes).